### PR TITLE
Allow to change RHS table in each join

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/40890
Demo https://www.loom.com/share/42ec7fdd09b34d09a75a3254a04c759f

Now we allow to change the rhs table for a join. If there are suggested join conditions, the join can be updated normally. If not, we store the rhs table temporary and enter our draft mode which was previously used for new join conditions only. When the join becomes complete, we replace the old join with the new one.

The only side effect / downside is that the old join remains in the query until there is at least one new join condition. So if you select another table and don't select the columns, then save or visualize - the query won't be changed.

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  join: Lib.Join;
+  color: string;
+  isReadOnly: boolean;
+  isModelDataSource: boolean;
+  onJoinChange: (newJoin: Lib.Join) => void;
+  onQueryChange: (newQuery: Lib.Query) => void;
+  onDraftRhsTableChange: (newTable: Lib.Joinable) => void;
+}
+
+export function JoinComplete({
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:28]: Moved from `Join`

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    () => Lib.displayInfo(query, stageIndex, rhsTable).displayName,
+    [query, stageIndex, rhsTable],
+  );
+
+  const handleStrategyChange = (newStrategy: Lib.JoinStrategy) => {
+    const newJoin = Lib.withJoinStrategy(join, newStrategy);
+    onJoinChange(newJoin);
+  };
+
+  const handleTableChange = (newTable: Lib.Joinable) => {
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:59]: Everything else is unchanged

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
+import { useMemo, useState } from "react";
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:1]: User needs to click the text (name of table) to edit RHS table despite `cursor: pointer` shown for the entire button:

https://github.com/metabase/metabase/assets/6830683/b6f695a9-234d-4a6a-ad36-ec54c219098a

**Review by kamilmielnik — COMMENTED** [review]
Looks good, left comments inline.

I also found #40940 while testing this.

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:1]: Good catch, I'll fix that

Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:1]: It's still there, I get it very often.

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      query,
+      stageIndex,
+      newTable,
+      joinPosition,
+    );
+    if (newConditions.length) {
+      const newJoin = Lib.joinClause(newTable, newConditions);
+      onJoinChange(newJoin);
+    } else {
+      onDraftRhsTableChange(newTable);
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:72]: Should be fixed now

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:1]: @kamilmielnik fixed. I did it in the same way as it is in DataStep

**Review by kamilmielnik — APPROVED** [review]
:+1:

**Pull request merged**